Gui

danielskeenan opened the pull request:
Fixes #1.

Summary of changes:

- Uses Qt 6.8 (minimum) to provide a graphical frontend.
  - Graphical version cannot be distributed as a single binary because of shared library requirements. It will need to be distributed as a zip/tarball/deb/rpm/installer etc. CPack should make this easy.
  - Uses same config file format as CLI just with a different loader.
- Added a few extra data members to DTOs for better presentation.
- Some classes (Options and Layout are the big ones) gained setters where they previously had none.
- Reworked solver output around callbacks instead of outputting directly.
- Lifted big chunks of README.md into the UI with light editing to suit context.

Thoughts?

beta382 commented on fnsolver/data/fnsite.h:
```diff
@@ -10,6 +10,8 @@
 #include <vector>
 
 struct FnSite {
+  friend bool operator==(const FnSite&, const FnSite&) = default;
```
There should only ever be one instance of a given FnSite (though I suppose I didn't make the constructor private), so equality check can be achieved more efficiently by address comparison.

Same goes for Probe (which I actually do a few times in the solver algorithm, e.g. chain resolution, so it is assumed that two identical probes are the same instance).

beta382 commented on fnsolver/layout/layout.cpp:
```diff
@@ -215,35 +236,43 @@ std::optional<Layout> Layout::from_frontier_nav_net_url(const std::string &url)
   return placements;
 }
 
+const Probe* Layout::get_probe(const FnSite& site) const {
```
I didn't document this, but it is assumed that `placements` (and consequently, `resolved_placements`) are complete and ordered according to site id (the resolution algorithm doesn't work otherwise). So you could do `layout.get_placements()[FnSite::idx_for_id.at(site.site_id)].get_probe()` (and if you want this function for shorthand, maybe return a reference for ease of use). Alternatively, you can go through `resolved_placements` (the first probe is guaranteed to be the site's placed probe).

beta382 commented on fnsolver/layout/layout.cpp:
```diff
@@ -215,35 +236,43 @@ std::optional<Layout> Layout::from_frontier_nav_net_url(const std::string &url)
+  return &placement->get_probe();
+}
+
+void Layout::set_probe(const FnSite& site, const Probe& probe) {
```
I don't think this is a good idea. For starters, I want Layout to be immutable. But aside from that, making modifications piecemeal is slow (since you have to redo the calculation each time). Not sure what you're using this for, but consider making a copy of `get_placements()`, doing your modifications to it, and creating a new layout from that.

danielskeenan replied:
This gets used in the map to allow manually changing a probe from the map, where only one is changed at a time. Since layouts are referenced in a few places (the map, inventory, and current yield displays), this felt like the best solution than copying the entire thing for a single change.

Although looking at this more, I think I can find some more effeciences in the map generation algorithms (as that's the only place `get_probe` is used) that don't rely on the binary search.

Will tweak some more.

beta382 commented on fnsolver/layout/layout.cpp:
```diff
@@ -215,35 +236,43 @@ std::optional<Layout> Layout::from_frontier_nav_net_url(const std::string &url)
   return resolved_placements;
 }
 
+const ResolvedPlacement& Layout::get_resolved_placement(const FnSite& site) const {
```
Same as with `get_probe()`

beta382 commented on fnsolver/solver/solver.h:
```diff
@@ -17,14 +17,25 @@
     Solver &operator=(Solver &&other) = delete;
 
-    Solution run(std::ostream &out) const;
+    Solution run(ProgressCallback progress_callback, StopCallback stop_callback) const;
```
Pass by const ref? Or `std::move` at the call sites.

beta382 commented on fnsolver/cli/main.cpp:
```diff
@@ -278,6 +279,8 @@ void output_options_report(const Options &options) {
 }
 } // namespace
 
+bool should_stop = false;
```
I'm not super familiar with C++'s memory visibility guarantees, but this should probably be an atomic wrapper.

danielskeenan replied:
This should definitely be atomic; it's in a signal handler! Whoops...

beta382 commented on fnsolver/cli/main.cpp:
```diff
@@ -319,7 +322,34 @@ int main(int argc, char **argv) {
+    std::cout << std::format("  Yield for best score:") << std::endl;
+    iteration_status.best_layout.output_report(std::cout, 4, false, true, false, false);
+  };
+  std::signal(SIGINT, [](int) { should_stop = true; });
```
I'm not going to be super pedantic about style (I cba tbh), but I'd super prefer if the callbacks were defined together, and there was some blank line separation before/after the signal handling + `solver.run`. Just too many lines all smushed together.

beta382 commented on fnsolver/solver/solver.cpp:
```diff
@@ -99,10 +99,7 @@ Solver::Solver(Options options)
       }()) {}
 
-Solution Solver::run(std::ostream &out) const {
-  should_stop = false;
```
`should_stop` is now unused, can delete on line 27

beta382 commented on CMakePresets.json:
```diff
@@ -21,7 +21,7 @@
     },
     {
-      "name": "release-win",
+      "name": "release-win-mingw",
```
Can you please update the Building section in the README accordingly (and include details about where one would get Qt / make it available for building)? Would be very much appreciated.

Also please add to the README links to the new third-party libraries you're using (maybe with a parenthetical that it's used in the GUI), remove the "- No GUI" bullet point, and feel free to (or to not) add a blurb about the GUI to the "Running" section.

danielskeenan commented:
I'll add some GitHub actions to do a build.

danielskeenan commented:
Wow, getting this to build in Actions was far more of a PITA than it usually is... I need to make some changes since I missed that the Qt version in Ubuntu LTS is a bit old (6.4). Not big changes, but not really trivial either.

[Here's a Windows build to try](https://github.com/danielskeenan/fnsolver/actions/runs/16896544917/artifacts/3740431296)

danielskeenan commented:
Artifacts on this run are up to date for Linux and Windows platforms - https://github.com/danielskeenan/fnsolver/actions/runs/16923233378

beta382 commented:
First off, thanks immensely for the github actions. I wasn't even aware github had built-in CI/CD (I'm used to Bitbucket+Jenkins).

A couple notes from my tests of the GUI (specifically using the tar.gz on ubuntu):

- Something is funky with the seed functionality (my understanding from using the UI and not looking at the code is that you use "everything in the current layout view less Basic Probes/None Probes" as the seed).
  - It's bypassing the inventory limits, I can reproduce this pretty consistently with the equivalent of `M5:4 M6:10 M7:4 M8:23 M9:10 M10:11 B1:6 B2:6 D:10 S:22`, Duplicators on 405, 408, 409 (basic probes everywhere else), 50 iterations, seed on. I would guess that you're not accounting for the seed in Options' `probe_quantities`. I have a few comments in options.h for what it expects of its data, `probe_quantities` needs "ordered, corrected for seed, filled with basic probes as needed". You can find how I handle this in cli_options.cpp (`adjust_probe_quantities_for_seed_and_fill`).
  - Force Seed doesn't seem to work. I can reproduce this with the above, except put the duplicators on 201, 206, 213. (I double checked that it works properly in CLI; the specified sites should have a Duplicator on them in the final solution). Not sure if this is related to the other seed issue, or the setting just isn't tied together.
- The solutions that get generated include basic probes when it shouldn't. With the same inventory from the example above, there should be 106 probes for 104 sites, and thus no basic probes. But I'm getting solutions that have many basic probes. Something must be going wrong with `probe_quantities` population.
- "Max Mining" in Tiebreaker can omit the bit about using Max Effective Mining, since it isn't relevant there
- Once the seed stuff is sorted, maybe add a bit to the alt text for "Seed" about how the seed is determined from your layout?
- Maybe a way to clear the displayed layout without clicking "New"? It's a bit annoying having to re-configure everything if you want to try a different seed. I suppose you could save off a minimal configuration and reload it, but I feel like this would be super convenient.

Separately, are you still considering my comments in layout.cpp? I am willing to compromise on a mutable Options, but I'd prefer if you found an alternative solution that leaves Layout immutable. The added getters as a shorthand are fine (though you can avoid the `find_if`, as I mentioned).

Finally, how do the build artifacts align with dynamically linking to qt? I didn't look too deeply into how you're building the GUI, nor did I look at the Windows artifacts (yet).

danielskeenan commented:
Re seeding - I thought I followed the same logic for setting up the solver as the CLI, but clearly I missed something. Will investigate. Might be related to the basic probes issue as well if `probe_quantities` isn't getting setup properly.

I also thought about adding an "empty layout" button as well, wanted to make sure the basics were good before working on icing.

Will rework Layout to keep it immutable.

The Linux packages either depend on Qt6 libraries from the distro's package manager (or assume they're already installed in the case of the tarball). The Windows installer installs Qt6 DLLs alongside the EXE.

danielskeenan commented:
The latest build should solve these issues. I also added some additional doc to README.

beta382 commented:
Seed and inventory stuff seem solved.

Something I just noticed (I don't believe this is a regression, I think I experienced it before but passed it off as me accidentally scrolling on the input area):

1. Configure a standard inventory
2. Run the solver (can be short) with an "All" Precious Resource constraint on any resource (e.g. Bonjelium).
3. See that the solution indeed has that resource at 100%
4. Open the solve configuration again, see that the Precious Resource constraint is still set to "All" visually
5. Rerun the solver with unchanged configuration
6. See that the solution does not have that resource at 100%
7. Open the solve configuration again, see that the Precious Resource constraint now is set to "Any"

Also happens if you make other changes between runs. I ran in to this when doing a test where I solved for a layout with a Precious Resource constraint, and then I used the resulting whole layout as an unforced seed for a second run (e.g. to try and deeply hone in on minor improvements, and also verify unforced seeding).

danielskeenan commented:
That was a puzzler! Looks like some internal state didn't get updated properly when the dialog was constructed and the none of the constraint widgets were interacted with.

beta382 commented:
Functionality lgtm. 

Ran the GUI through a handful of scenarios, all seemed to work as expected. Ran on Linux (ubuntu) + Windows with the Actions-built zip/tarball artifacts (didn't try the packages, just trusting them).

One thing I am concerned about is that the Windows CLI is substantially slower (like 40-50% more time) than the one on `main` (Linux CLI is the same speed as `main`, GUIs are same speed as CLI on their respective platforms). Is MSVC just that much worse at optimization here than mingw-gcc? Static linking difference for stdlib and whatnot? `<thread>` impl difference? Something with the build environment? You're using `/O2` and whole program optimization is enabled, which seems right. Not sure what's going on here.

If all fails, the CLI could be built separately by hand with mingw-gcc and distributed separately, but it would be nice for the GUI to have speed parity. It looks like mingw has a Qt build (https://packages.msys2.org/groups/mingw-w64-x86_64-qt6), but that's out of my depth.

I will say that I would like the `release-win-mingw` preset to actually work on linux builds (mvp: exclude the gui target). The other linux presets work fine with the GUI build (after I installed the relevant Qt packages).

---

I have a few "nice to have's" for you to yea or nay:

- Widen the inventory count column a bit by default (it's impossible to see what you're inputting when the up/down arrows pop in at the default width)
- A way to set the inventory to the equivalent of "all de" or "all og" in one click (maybe another menu bar entry); optionally with "remove all mining" and "remove all research" (I have others in the CLI for completeness but those are the ones that are actually time savers)
- Display of scores higher than 1M uses scientific notation. Not inherently an issue, but it *is* possible to naturally hit >1M revenue scores and it would be nice if they displayed fully. I don't really care about how very large Weights scores are displayed.
- It would be neat if the run progress dialogue, when running with bonus iterations, said "X iteration(s) remaining" once the standard iterations are exceeded (currently it just says "remaining"). I believe enough information is already present in IterationStatus for this. Of course, this will be an uncertain and jumpy measure, but I think it's better than nothing if you've been displaying a remaining expectation previously.
- Disallow tiebreaker when it's equal to score function (CLI parity). Maybe also default to None (force to None if made conflicting by a score function selection change).
- CLI chokes on the newlines inside arrays in the GUI-generated config file (specifically Inventory and Seed, maybe others if they exist but I didn't test too deep). If it's easy to configure toml++ to just write these to one line, that would be neat, otherwise I don't consider this a priority (ostensibly a CLI11 bug; it's specifically if the first line contains the open bracket and no elements, and it looks like they simply missed the "first line has just the open bracket and no elements" case when adding multi-line support in the past).

As a note for posterity, it's a bit awkward that saving off options includes a solved layout as the seed. It makes perfect sense for the GUI (because "enable seed" is a separate option and "seed" captures the displayed layout), but the CLI will not like it (it works correctly, but the seed is on if it's present and that's a bit unexpected when its "an entire layout"). I'm willing to file this under "this behavior is correct", "this will probably affect 0 people", and "I'll investigate a workaround on the CLI side iff someone complains".

danielskeenan commented:
I've never actually tried the mingw build of Qt, mostly because I've always targeted whatever the "native" compiler is on a given platform. Will give it a go. At the very least, it's easy enough to leave out building the GUI if Qt can't be found in the search path. Would definitely be interesting to benchmark the results of the two compilers. I don't *think* I changed much in the business logic that would affect performance that greatly; the most I can think of is that the progress callback involves a copy to get around threading issues, but I don't think that struct should be that expensive to copy.

I think most of the icing is possible, but I could use some clarity on bonus iterations (see `RunProgressDialog::progress` for how it's displayed currently). Are you suggesting that the displayed max iterations change if we go into bonus iterations?

I was toying around with having an "OG mode" and a "DE mode" that not only allowed setting the inventory to the max available, but also added a column to the inventory showing that maximum so a user can track their collection progress (i.e. "stop looking for more mining G10, you've found them all) and changed the graphics. I haven't found a good source of probe icons in the DE yet, but can probably do the rest without the graphics.

Re. the config file issue - There's some [format flags](https://marzer.github.io/tomlplusplus/namespacetoml.html#a2102aa80bc57783d96180f36e1f64f6a) that can be applied, but none of them seem applicable here. I also feel like, if someone is savvy enough to use the CLI over the graphical option, they're (hopefully) savvy enough to remove the seed from their config file to get the CLI to do what they want. Not sure where the best place to document this is though.

beta382 commented:
> I've never actually tried the mingw build of Qt, mostly because I've always targeted whatever the "native" compiler is on a given platform. Will give it a go. At the very least, it's easy enough to leave out building the GUI if Qt can't be found in the search path. Would definitely be interesting to benchmark the results of the two compilers. I don't _think_ I changed much in the business logic that would affect performance that greatly; the most I can think of is that the progress callback involves a copy to get around threading issues, but I don't think that struct should be that expensive to copy.

It's definitely nothing to do with the code itself. On Linux, it runs exactly the same speed as before. It's just on Windows that it's slower. Maybe the stdlib it's dynamically linking against isn't built with optimizations? Idk. It's something with the build configuration or run environment. I have trouble believing the compilers are that far apart on optimization.

> I think most of the icing is possible, but I could use some clarity on bonus iterations (see `RunProgressDialog::progress` for how it's displayed currently). Are you suggesting that the displayed max iterations change if we go into bonus iterations?

Not the max iterations (I think `128/100` is fine, that's what I do with the CLI). The `time_remaining`. Once you hit the normal max iteration, it goes from (don't have it up in front of me, paraphrasing) "00:00:01 remaining" to " remaining". It would be neat if it started saying e.g. "22 iteration(s) remaining" (if you're at "last improvement: 28 iteration(s) ago" with bonus iterations = 50)

> I haven't found a good source of probe icons in the DE yet, but can probably do the rest without the graphics.

Asking around, I'll let you know if anyone bites.

> Re. the config file issue - There's some [format flags](https://marzer.github.io/tomlplusplus/namespacetoml.html#a2102aa80bc57783d96180f36e1f64f6a) that can be applied, but none of them seem applicable here. I also feel like, if someone is savvy enough to use the CLI over the graphical option, they're (hopefully) savvy enough to remove the seed from their config file to get the CLI to do what they want. Not sure where the best place to document this is though.

Not worth pursuing then. If someone complains, I'll look into it. I'll file a bug with CLI11 for the line-wrapped array parsing thing.

danielskeenan commented:
Oh, that `XX remaining` thing is a bug then. There's probably an underflow somewhere since the iteration number will have exceeded the max from options and it's throwing off the time estimate.

beta382 commented:
Right.

Once you're in bonus iterations territory though, I don't think a time estimate is valuable any longer. It'll end once you stop getting improvements, which isn't really estimable. If you have a better idea than printing the current number of remaining non-improved iterations, go for it.

danielskeenan commented:
To rephrase the requirement:
- When `iteration <= options.max_iterations`, display a time estimate.
- When `iteration > options.max_iterations`, display `options.bonus_iterations - last_improvement` and change the progress bar to a busy indicator (that thing where the bar scrolls infinitely rather than showing actual progress, otherwise it will appear stuck at 100%).

beta382 commented:
`options.bonus_iterations - (iteration - last_improvement)`, but yes.

danielskeenan commented:
> Display of scores higher than 1M uses scientific notation. Not inherently an issue, but it is possible to naturally hit >1M revenue scores and it would be nice if they displayed fully. I don't really care about how very large Weights scores are displayed.

Where are you seeing this? I forced some big numbers in various places and cannot reproduce.
![](https://github.com/user-attachments/assets/d22abd58-37dc-4009-9051-ce704c83dea2)
![](https://github.com/user-attachments/assets/761fe741-488f-4084-8305-1a43016c54cc)

beta382 commented:
<img width="410" height="326" alt="image" src="https://github.com/user-attachments/assets/463b6b3f-adbc-4527-b1a3-81f25a07ecc6" />

Both building manually and the latest artifact on your fork, Ubuntu 24.04.2 (VM on Win 10 host). It shows comma-grouped until 7 digits. Also ran the latest artifact on your fork on Win 10 (side note, is the map and app in general supposed to have a white background on Windows, see below. I have the system setting on dark mode for apps).

Displays as expected in the Results panel.

Ran with default config, 6 of each booster, 10 duplicators, 12 Research G6, max revenue.

I manually widened the modal to make sure it wasn't a space constraint automatic thing.

<img width="1005" height="789" alt="image" src="https://github.com/user-attachments/assets/3bfe7279-44a4-42ff-a9a7-1ad0a39191d6" />

danielskeenan commented:
Give [this run](https://github.com/danielskeenan/fnsolver/actions/runs/17023626701) a shot!

beta382 commented:
Punch list looks mostly good.

I've seen a couple brief moments where `time_remaining` still says " remaining" during the transition, but I don't think you're off-by-one or anything (I did a run with very high population and offspring to give really slow iterations, and it went from time to bonus iterations without a gap just fine). I wager that occasionally your `time_required - time_elapsed` goes negative, and that's invalid, so it formats to an empty string. Could probably just catch and default to "00:00:00" (or set invalid time to a zero time, if that's really it).

When tiebreaker is set to None, a subsequent run has no box checked in the configuration. Couldn't really tell what the actual behavior is when running a subsequent run without re-checking anything (I'd assume still None; it doesn't break, at least).

---

Windows GUI is *much* closer to speed parity, and CLI is all the way there (see below). I guess MSVC really is that far behind on the specific optimizations this benefits from? Appreciate you pursuing this.

Windows CLI needs mingw installed or DLLs adjacent (specifically `libgcc_s_seh-1.dll`, `libstdc++-6.dll`, and `libwinpthread-1.dll`). If it's not too much trouble, building the CLI fully statically linked would be nice for ultimate ease (I did this from the Preset before, but with multiple executable targets now I think it'll have to be a conditional in the cli cmake instructions), but otherwise packaging the required DLLs adjacent would work.

Running the Github Actions artifacts with the equivalent of `-f max_effective_mining 2 -i all_de no_research M1:0 M2:0 M3:0 M4:0 M5:4 -n 200 -t 6 -y` on my hardware:

Platform|CLI `main`|CLI PR|GUI PR
-|-|-|-
Ubuntu 24.04.2 VM|20.81|20.71|20.60
Windows 10|21.40|21.69|23.29

CLI timed with `time` in bash and `Measure-Command` in PowerShell, with ye olde manual stopwatch for the GUI. Three-run average for each cell.

---

What are the prereqs for building `release-win-mingw` now? I could follow the `apt-get` package set in the pipeline to get the standard release build building, but I'm not sure what I need to install or configure to cross-compile.

danielskeenan commented:
> I've seen a couple brief moments where time_remaining still says " remaining" during the transition, but I don't think you're off-by-one or anything (I did a run with very high population and offspring to give really slow iterations, and it went from time to bonus iterations without a gap just fine).

Added some `std::max` to stop it going negative.

> When tiebreaker is set to None, a subsequent run has no box checked in the configuration. Couldn't really tell what the actual behavior is when running a subsequent run without re-checking anything (I'd assume still None; it doesn't break, at least).

Fixed, there was a conditional that didn't need to be there after re-working how score functions were serialized/copied/compared to use the same path.

> Windows GUI is much closer to speed parity, and CLI is all the way there (see below). I guess MSVC really is that far behind on the specific optimizations this benefits from? Appreciate you pursuing this.

I was floored by the difference! (I usually use Fedora, so didn't catch the performance issues before you pointed them out) I added a `debug-win-mingw` cmake preset for parity between platforms. Also adjusted CMake to statically link the CLI to the mingw runtime; the GUI gets the DLLs deployed alongside.

I have a theory about your light/dark mode issues; I'm waiting for a Windows 10 VM to spin up to check. Basically Qt uses special platform plugins to make the application look "native" and I suspect some differences in the light/dark detection between Windows 11 and Windows 10.

> What are the prereqs for building release-win-mingw now?

If you ask The Qt Group, you'll need to [get the official install Qt](https://www.qt.io/download-qt-installer-oss), choose the mingw variant, not the msvc variant, as well as a matching version of mingw (under tools).

CI goes a different route and uses [`aqt`](https://aqtinstall.readthedocs.io/en/latest/getting_started.html) which I think is a little easier.
```
aqt install-qt windows desktop 6.8 win64_mingw --autodesktop --outputdir <WHERE YOU WANT IT>
aqt install-tool windows desktop tools_mingw1310 --outputdir <SAME DIR AS ABOVE>
```

Once installed, add the install directory to `CMAKE_PREFIX_PATH`. Because of the shared library issue, running the program also requires it be installed rather than only built. There is an option called `BUILD_PACKAGE` that, when enabled, creates the targets that build the installers. My full cmake command line on Windows is
```
cmake
--preset "release-win-mingw"
-DCMAKE_PREFIX_PATH=<WHERE YOU PUT IT>/6.8.3/mingw_64
-DBUILD_PACKAGE=On
-DCMAKE_INSTALL_PREFIX=./build-release-win/install_root

cmake --build ./build-release-win

cmake --install ./build-release-win
```
Which will, once completed, put binaries in build-release-win/install_root/bin

beta382 commented:
> I have a theory about your light/dark mode issues; I'm waiting for a Windows 10 VM to spin up to check. Basically Qt uses special platform plugins to make the application look "native" and I suspect some differences in the light/dark detection between Windows 11 and Windows 10.

Not a dealbreaker, just was curious.

---

All functionality looks good, tested the static linked Win CLI without libs present and it worked fine. If you could touch up the README Building section with the updated presets (maybe a blurb about the GUI build being skipped if you don't have Qt), and any other nuances you think are worth mentioning, that would be great.

Beyond that, I'm willing to approve this if there's nothing else you want to add or change. Let me know.

danielskeenan commented:
Will update the readme. My hunch was correct, dark mode should work in the new build!

danielskeenan commented:
Alright, I think it's ready to go. I changed the mingw toolchain file to autolocate Qt if it's installed with aqt per the instructions which should make it easier.

danielskeenan commented:
GitHub pull requests can only be merged by someone with commit access to the main repo, so that's up to you!

Since you mentioned being new to GitHub Actions, if it's something you're into, [here's](https://github.com/danielskeenan/mobile_sacn/blob/7f8189772748f677c28c59097e79f26befb60ac4/.github/workflows/main.yml#L141) how I've setup another project of mine to add the built artifacts automatically to a draft release when a commit is tagged. It won't make the release public until you click the Big Button, but reduces the tedium of combining the different platform artifact zips together and uploading them to a release.

beta382 commented:
Ahh, as you can tell I am not accustomed to github adminstration. Thanks again!
